agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,345 @@
+  }
+  hashmap = *hashmap_ptr;
+
+  stats.elements = hashmap->elements;
+  stats.collisions_min = stats.elements;
+
+  count = nr_count_buckets(hashmap);
+  for (i = 0; i < count; i++) {
+    nr_wraprecs_bucket_t* bucket = hashmap->buckets[i];
+    bucket_items_cnt = 0;
+
+    if (bucket) {
+      stats.buckets_used++;
+    }
+
+    while (bucket) {
+      nr_wraprecs_bucket_t* next = bucket->next;
+
+      bucket_items_cnt++;
+
+      nr_destroy_wraprecs_bucket(&bucket, hashmap->dtor_func);
+      bucket = next;
+    }
+
+    if (0 != bucket_items_cnt && bucket_items_cnt < stats.collisions_min) {
+      stats.collisions_min = bucket_items_cnt;
+    }
+    if (bucket_items_cnt > stats.collisions_max) {
+      stats.collisions_max = bucket_items_cnt;
+    }
+    if (bucket_items_cnt > 1) {
+      stats.buckets_with_collisions++;
+      stats.collisions_mean += bucket_items_cnt;
+    }
+  }
+
+  if (0 != stats.buckets_with_collisions) {
+    stats.collisions_mean /= stats.buckets_with_collisions;
+  }
+
+  nr_free(hashmap->buckets);
+  nr_realfree((void**)hashmap_ptr);
+
+  return stats;
+}
+
+static bool nr_zf_is_unnamed_closure(const zend_function* zf) {
+  if (9 != ZSTR_LEN(zf->op_array.function_name)) {
+    return false;
+  }
+  return 0 == memcmp("{closure}", ZSTR_VAL(zf->op_array.function_name), 9);
+}
+
+void nr_php_wraprec_hashmap_key_set(nr_php_wraprec_hashmap_key_t* key,
+                                    const zend_function* zf) {
+  key->lineno = nr_php_zend_function_lineno(zf);
+  key->scope_name = NULL;
+  key->function_name = NULL;
+  key->filename = NULL;
+
+  if (NULL != zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    key->function_name = zf->op_array.function_name;
+    zend_string_addref(key->function_name);
+    if (zf->op_array.scope) {
+      key->scope_name = zf->op_array.scope->name;
+      zend_string_addref(key->scope_name);
+    }
+  } else {
+    if (NULL != zf->op_array.filename) {
+      key->filename = zf->op_array.filename;
+      zend_string_addref(key->filename);
+    }
+  }
+}
+
+void nr_php_wraprec_hashmap_key_release(nr_php_wraprec_hashmap_key_t* key) {
+  if (NULL != key->scope_name) {
+    zend_string_release(key->scope_name);
+  }
+  if (NULL != key->function_name) {
+    zend_string_release(key->function_name);
+  }
+  if (NULL != key->filename) {
+    zend_string_release(key->filename);
+  }
+  key->lineno = 0;
+}
+
+static inline size_t nr_zendfunc2bucketidx(size_t log2_num_buckets,
+                                           zend_function* zf) {
+  /* default to lineno */
+  uint32_t hash = nr_php_zend_function_lineno(zf);
+
+  if (NULL != zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    /* but use hash of function name when possible */
+    hash = ZSTR_HASH(zf->op_array.function_name);
+  } else if (NULL != zf->op_array.filename) {
+    /* and as last resort try hash of filename if available */
+    hash = ZSTR_HASH(zf->op_array.filename);
+  }
```

Comment:
my hash implementation knowledge is rusty, but couldn't we hypothetically increase hash variability and reduce collisions if we added the hash values rather than just setting them based on an individual value? 
For example, 
lineno hash + function_name hash should be more unique than just lineno or just function name?

Reply:
Not really. Actually since lineno has very little entropy combining it with function_name or filename hash may reduce entropy from function_name hash especially that Zend's engine hashing function seems to be pretty well designed for strings and function names tend to be fairly unique (IMHO more unique than line numbers). Additionally the key comparison in this hashmap compares either function_name or filename so combining these two hashes also doesn't seem to make sense. I guess we'll have to wait for collisions stats from the field.